block/manager_test.go

```diff
@@ -94,3 +95,88 @@ func TestIsDAIncluded(t *testing.T) {
 	m.blockCache.setDAIncluded(hash.String())
 	require.True(m.IsDAIncluded(hash))
 }
+
+func TestSubmitBlocks(t *testing.T) {
+	require := require.New(t)
+	ctx := context.Background()
+
+	m := getManager(t)
+
+	maxDABlobSizeLimit, err := m.dalc.DA.MaxBlobSize(ctx)
+	require.NoError(err)
+
+	testCases := []struct {
+		name                        string
+		blocks                      []*types.Block
+		isErrExpected               bool
+		expectedPendingBlocksLength int
+	}{
+		{
+			name:                        "happy path, all blocks A, B, C are submitted on first round",
+			blocks:                      []*types.Block{types.GetRandomBlock(1, 5), types.GetRandomBlock(2, 5), types.GetRandomBlock(3, 5)},
+			isErrExpected:               false,
+			expectedPendingBlocksLength: 0,
+		},
+		{
+			name: "blocks A and B are submitted first round because including c triggers blob size limit. C is submitted on second round",
+			blocks: func() []*types.Block {
+				// Find three blocks where two of them are under blob size limit
+				// but adding the third one exceeds the blob size limit
+				var block1, block2, block3 *types.Block
+				for numTxs := 0; ; numTxs += 100 {
+					block1 = types.GetRandomBlock(1, numTxs)
+					blob1, err := block1.MarshalBinary()
+					require.NoError(err)
+
+					block2 = types.GetRandomBlock(2, numTxs)
+					blob2, err := block2.MarshalBinary()
+					require.NoError(err)
+
+					block3 = types.GetRandomBlock(3, numTxs)
+					blob3, err := block3.MarshalBinary()
+					require.NoError(err)
+
+					if uint64(len(blob1)+len(blob2)) < maxDABlobSizeLimit && uint64(len(blob1)+len(blob2)+len(blob3)) > maxDABlobSizeLimit {
+						return []*types.Block{block1, block2, block3}
+					}
+				}
+			}(),
+			isErrExpected:               false,
+			expectedPendingBlocksLength: 0,
+		},
+		{
+			name: "A and B are submitted successfully but C is too big on its own, so C never gets submitted",
+			blocks: func() []*types.Block {
+				numBlocks, numTxs := 3, 5
+				blocks := make([]*types.Block, numBlocks)
+				for i := 0; i < numBlocks-1; i++ {
+					blocks[i] = types.GetRandomBlock(uint64(i+1), numTxs)
+				}
+				for numTxs := 0; ; numTxs += 100 {
+					block3 := types.GetRandomBlock(3, numTxs)
+					blob3, err := block3.MarshalBinary()
+					require.NoError(err)
+
+					if uint64(len(blob3)) > maxDABlobSizeLimit {
+						blocks[2] = block3
+						return blocks
+					}
+				}
```

Comment:
[non blocking]

this could be a helper I think.  have a function that accepts a target block size and returns a block of that size. Then we could target blocks of size 0, max / 2, max.

Reply:
Added